### ISSUE TITLE
3.9.0/Ubuntu: Add manual connection refresh feature

### ISSUE THREAD

**Issue by erezhdd** [sidneys/pb-for-desktop#21]
PB4D 3.9.0 / Ubuntu 16.04.2:
When the laptop is resumed, or when network connectivity is renewed after being disconnected, PB stops reporting notifications. Signing out and then back in seems to resolve this. Can you please have the app reset the connection automatically whenever it detects network connectivity being resumed?

**Comment by erezhdd** [comment]
I have another idea for an in-between solution that is much simpler: can you implement a "Refresh PB connection" action in the tray menu that will reset the connection to PB? This way, the detection of connectivity issues can be delayed for later, and the tray action should also cover for possible future bugs in the detection.

**Comment by sidneys** [comment]
@erezhdd 

Thanks for your suggestion. Can you please add additional info about the reconnection issue, e.g. how did the disconnect occur?

**Comment by erezhdd** [comment]
Quite simple actually. My desktop is a laptop, which I often take away for meetings, work on the bus etc. Consequently, the network often disconnects and reconnects. Aside from that, I frquently suspend and resume the laptop, which has a similar network effect. When reconnecting, pb app typically stops notifications. To reset it, I usually have to open it, click sign out and sign in again. In some cases the main window also blackens, so I have to quit the app and start it again. Having the suggested reset action should ease this problems considerably.

**Comment by sidneys** [comment]
@erezhdd 

Could not reproduce this issue at all on a Ubuntu 16.10 Virtual Machine.

After disconnecting the network or suspend-and-resume, the app picks up the notifications again after a short moment. 

Can you interact with the application after resuming, is the webapp completely non-responsive, or is it just the notifications which are missing?

**Comment by erezhdd** [comment]
I seriously doubt the test equivalence of a VM using a virtual network adapter and more advanced software. 

Per your request: Last night I saw that notifications resume after changing network connection (wired->wifi),but not when using connections such as phone tethering (USB or bluetooth), and not over suspend and resume. Right now, with my machine resumed to use wifi in a meeting room, I don't receive any notifications for an ongoing whatsapp discussion (even though I see the notifications on the phone). 

I would really like the option to reset the PB connection.

**Comment by sidneys** [comment]
@erezhdd 

Thanks for the additional details, appreciate it. As for the suspend/resume, there is now an auto-reload feature in [v3.9.96](../releases) when resuming from laptop sleep. Please review and post whether this this solves that part of the issue.

A "reload" button should be the last resort. Automatic connection dropout handling is the preferred solution 😎.

**Comment by erezhdd** [comment]
I'm glad you strive for quality, delivering value at minimum user interference. However, there are likely going to be issues with detecting the various conditions. Wouldn't you prefer to have this button for "just in case"? Note that you already  implemented the core function (restart) in 3.9.96. All that remains is to connect it to the GUI.
I will update the version, and report back. Thanks for being so responsive.

**Comment by erezhdd** [comment]
Initial tests seem to work past resume. Two points:
- Starting the app first time now dumps an error message. See attached.
- I hope you capture the network resume event and not the machine resume. Network resume happens a little later than machine resume and that's the defining event.
![screenshot from 2017-03-15 19-15-52](https://cloud.githubusercontent.com/assets/1337495/23962168/e3ae35ea-09b5-11e7-9ba0-89adf2dfe9f2.png)

**Comment by sidneys** [comment]
Hi @erezhdd 

There is now a [new version](../releases) with logout as well as reconnect interface. Please let us know whether this solves the issue, as well as the issues which have occurred intermittently.

**Comment by erezhdd** [comment]
First, the GUI seems a lot more usable. I'd skip the confirmation dialogs (or at least allow to skip them with checkbox or configuration). This is because PushBullet is a stateful service - there is no loss of data upon disconnection. Also, the error message from my previous comment is gone. Only "PB for desktop upgraded to 3.99" appears at first run.
The main issue - **It's not connecting anymore**. Below is the constant main window look. I tried reconnecting, disconnecting, quitting the app and running it again. Nothing.
![screenshot from 2017-03-16 12-02-43](https://cloud.githubusercontent.com/assets/1337495/23991238/46304ede-0a41-11e7-9f23-9c1434ca9bed.png)

**Comment by sidneys** [comment]
@erezhdd 
Thanks for the follow up and screencaps. Please retry with the new [v4.0.2](../releases) and report your findings.

**Comment by erezhdd** [comment]
App now connects ok. Testing the reconnection will take some more time.

**Comment by sidneys** [comment]
@erezhdd
Version [4.1.0](../releases/) was just released as well, further improving on a lower resource footprint.

**Comment by erezhdd** [comment]
4.1.0: Did the upgrade. Basic connection works out-of-the-box. Reconnect button proving to be quite valuable. I resumed my laptop over 30 minutes ago. I heard my phone receiving notification but nothing showed on my desktop screen. I clicked reconnect and the next notification appeared properly on the screen. That's exactly what I meant by having this button - covering for detection issues. There could also be bugs in the OS or configuration issues that interfere with detection, out of the app control, so this reconnect button is very handy. Thanks!
